Chess/state.h

```diff
@@ -0,0 +1,87 @@
+#ifndef STATE_H
+#define STATE_H
+
+#include "common_types.h"
+#include "game.h"
+
+class Game;
+
+class State
+{
+public:
+    State();
+    virtual ~State();
+
+    virtual void cellAction(Game*, QString);
+
+    virtual void startAction(Game*);
+    virtual void stopAction(Game*);
+
+    virtual void saveAction(Game*, QString);
+    virtual bool loadAction(Game*, QString);
+
+    virtual void rollback_from_list(Game*);
+    virtual void make_move_from_list(Game*);
+
+protected:
+    void ChangeState (Game*, State*);
+
+};
+
+
+/////////////////////////////////////////////////////////////////////////
+class State1 : public State
+{
+public:
+    State1() {}
+
+    static State* Instance();
+
+    void cellAction(Game*, QString);
+    void startAction(Game*);
+    bool loadAction(Game*, QString);
+
+private:
+    static State *self_;
+
+};
+
+/////////////////////////////////////////////////////////////////////////
+class State2 : public State
+{
+public:
+    State2() {}
+
+    static State* Instance();
```

Comment:
regarding singltons:
1. one tiny thing is missing - constuctor is still avalible, we can create more than 1 instance of this class :)
2. prefere using "mayers singltone" - the one that is implemented via local static "self" value
3. singleton is an anti pattern in multithreaded environment, restrict their usage as much as possible.
However in this case singleton usage is justified, nevertheless I would suggest to instantiate them before program execution starts